qiskit/algorithms/phase_estimators/phase_estimation.py

```diff
@@ -0,0 +1,211 @@
+
+        How the frequencies are computed depends on whether the backend computes amplitude or
+        samples outcomes.
+
+        1) If the backend is a statevector simulator, then the reduced density matrix of the
+        phase-reading register is computed from the combined phase-reading- and input-state
+        registers. The elements of the diagonal :math:`(i, i)` give the probability to measure the
+        each of the states `i`. The index `i` expressed as a binary integer with the LSB rightmost
+        gives the state of the phase-reading register with the LSB leftmost when interpreted as a
+        phase. In order to maintain the compact representation, the phases are maintained as decimal
+        integers.  They may be converted to other forms via the results object,
+        `PhaseEstimationResult` or `HamiltonianPhaseEstimationResult`.
+
+         2) If the backend samples bitstrings, then the counts are first retrieved as a dict.  The
+        binary strings (the keys) are then reversed so that the LSB is rightmost and the counts are
+        converted to frequencies. Then the keys are sorted according to increasing phase, so that
+        they can be easily understood when displaying or plotting a histogram.
+
+        Args:
+            circuit_result: the result object returned by the backend that ran the QPE circuit.
+
+        Returns:
+               Either a dict or numpy.ndarray representing the frequencies of the phases.
+
+        """
+        if self._quantum_instance.is_statevector:
+            state_vec = circuit_result.get_statevector()
+            evaluation_density_matrix = qiskit.quantum_info.partial_trace(
+                state_vec,
+                range(self._num_evaluation_qubits,
+                      self._num_evaluation_qubits + self._num_unitary_qubits)
+            )
+            phases = evaluation_density_matrix.probabilities()
+        else:
+            # return counts with keys sorted numerically
+            num_shots = circuit_result.results[0].shots
+            counts = circuit_result.get_counts()
+            phases = {k[::-1]: counts[k] / num_shots for k in counts.keys()}
+            phases = _sort_phases(phases)
+            phases = qiskit.result.Counts(phases, memory_slots=counts.memory_slots,
+                                          creg_sizes=counts.creg_sizes)
+
+        return phases
+
+    def estimate(self,
+                 num_evaluation_qubits: Optional[int] = None,
+                 unitary: Optional[QuantumCircuit] = None,
+                 state_preparation: Optional[QuantumCircuit] = None,
+                 pe_circuit: Optional[QuantumCircuit] = None,
+                 num_unitary_qubits: Optional[int] = None) -> PhaseEstimationResult:
+
+        super().estimate(num_evaluation_qubits, unitary, state_preparation,
+                         pe_circuit, num_unitary_qubits)
+        return self._run()
```

Comment:
Yes

Reply:
The `_run` class in `HamiltonianPhaseEstimation` is different. I put all common code into `estimate` and renamed the `_run` method to `_result` because all it does now is instantiate the result class.